R/arc-select.R

```diff
@@ -454,10 +450,91 @@ count_results <- function(req, query) {
   resp <- httr2::resp_body_string(
     httr2::req_perform(
       httr2::req_url_query(n_req, f = "json"),
-      error_call = rlang::caller_env()
+      error_call = error_call
     )
   )
 
   RcppSimdJson::fparse(resp)[["count"]]
 }
 
+#' Set page size and check page size validity
+#'
+#' @noRd
+set_page_size <- function(
+    page_size = NULL,
+    max_records = 25000L,
+    error_call = rlang::caller_env()
+    ) {
+  # if its null, just use max records (default)
+  if (is.null(page_size)) {
+    return(max_records)
+  }
+
+  # ensure its an integer.
+  page_size <- as.integer(page_size)
+  page_size_len <- length(page_size)
+
+  if (page_size_len > 1) {
+    cli::cli_abort(
+      "{.arg page_size} must be length 1 not {page_size_len}",
+      call = error_call
+    )
+  }
+
+  if (page_size < 1) {
+    cli::cli_abort(
+      "{.arg page_size} must be a positive integer.",
+      call = error_call
+    )
+  }
+
+  if (page_size > max_records) {
+    cli::cli_abort(
+      "{.arg page_size} ({page_size}) cannot excede layer's {.field maxRecordCount} property ({max_records})",
+      call = error_call
+    )
+  }
+
+  page_size
+}
+
+
+
+#' Bind list with parsed ESRI geometry into sf object
+#'
+#' @noRd
+rbind_results <- function(res, crs = NULL, error_call = rlang::caller_env()) {
+  empty_res <- vapply(res, rlang::is_empty, FALSE)
+
+  if (all(empty_res)) {
+    cli::cli_alert_info("No features returned from query")
+    return(data.frame())
+  }
+
+  # Drop empty results from sf list
+  if (any(empty_res)) {
+    res <- res[!empty_res]
+  }
+
+  res <- rbind_res_list(res, error_call = error_call)
+
+  if (inherits(res, "sf") && is.na(sf::st_crs(res))) {
+    sf::st_crs(res) <- crs
+  }
+
+  res
+}
+
+#' Combine data frame list
+#'
+#' Inspired by arcpbf https://github.com/R-ArcGIS/arcpbf/blob/main/R/post-process.R
+#' @noRd
+rbind_res_list <- function(x, error_call = rlang::caller_env()) {
+  # TODO: enhance this with additional suggested packages
+  if (rlang::is_installed("vctrs")) {
+    x <- vctrs::vec_rbind(!!!x, error_call = error_call)
+  } else {
+    x <- do.call(rbind.data.frame, x)
+  }
+  x
+}
```

Comment:
It might be worth extracting this into arcgisutils 
https://github.com/R-ArcGIS/arcpbf/blob/main/R/post-process.R#L109-L121

```r
squish_dfs <- function(x) {
  if (rlang::is_installed("collapse", version = "2.0.0")) {
    x <- collapse::rowbind(x)
  } else if (rlang::is_installed("data.table")) {
    x <- data.table::rbindlist(x)
    data.table::setDF(x)
  } else if (rlang::is_installed("dplyr")) {
    x <- dplyr::bind_rows(x)
  } else {
    x <- do.call(rbind.data.frame, x)
  }
  x
}
```

Reply:
Is the speed boost for including data.table or collapse worth the added depencies? I'm mostly agnostic but didn't want to add a bunch of new dependencies without some clear confirmation.

FWIW – I've also been avoiding `dplyr::bind_rows()` ever since `purrr::map_dfr()` was superseded (and cited the "confusing semantics with edge cases" for `bind_rows` the updated description).